docs/doc/60-contributing/03-rfcs/20220729-recluster.md

```diff
@@ -0,0 +1,73 @@
+
+```
+pub struct ClusterStatistics {
+    ... ...
+    pub level: i32,
+}
+```
+
+The workflow of a recluster operation is divided into two tasks, block selection and block merge.
+
+### syntax
+
+```sql
+optimize table tbl_name [final] recluster 
+```
+
+If specify `final`, optimization is performed until the table is well clustered enough. Otherwise, the recluster workflow will only run once.
+
+The optimize statement should be triggered by DML on the table.
+
+### Metrics
+
+- overlap
+  The number of blocks that overlap with a given block.
+
+- depth
+  The number of blocks that overlap at the same point. The points are collected from the minimum and maximum value in the clustering values domain range.
+
+### Block Selection
+
+The initial level of newly incoming data is 0. We focus on the newer data first, in other words the selection operations are preferentially performed on level 0. The advantage of doing this is to reduce write amplification.
+
+1. Calculate the depth of each point and the overlaps of the block, and summarize to get avg_depth. The algorithm has already been reflected in [clustering_information.rs](https://github.com/datafuselabs/databend/blob/main/query/src/storages/fuse/table_functions/clustering_informations/clustering_information.rs#L109-L191), and will not be repeated here. The ideal result for avg_depth is 1. In order to achieve roughly ordering, consider defining a threshold or a ratio (threshold = blocks_num * ratio). As long as avg_depth is not greater than this threshold, the blocks at this level can be considered well-clustered, then we perform block selection on the next level.
+
+2. Select the point range (one or more) with the highest depth, and select the blocks covered by the range as a set of objects for the next block-merge. If there is more than one range with the highest depth, there may be multiple sets of blocks that can be parallelized during block-merge.
+
+Tip:
+```
+1. The cluster key may be created or altered when the table has data, so there may be blocks that are not sorted according to the cluster key. Consider temporarily ignoring such blocks when doing recluster.
+
+2. If the cluster key of a block has only one value (the maximum and minimum values are equal, reaching the constant state) and its row_num is 1000_000, set its level to -1 and filter it out when doing recluster.
+
+3. The selected blocks maybe need to consider the total size, otherwise the sorting may be out of memory.
+```
+
+### Block Merge
+
+Sort and merge the collected blocks. After the merged block exceeds a certain threshold (1000_000 rows), it will be divided into multiple blocks. The newly generated block is put into the next level.
```

Comment:
How large those blocks could be in storage?

Reply:
The capacity of blocks depends on the size of the memory. Currently, sort does not support external sort. Large-capacity blocks may cause oom.